Feat/fill form snippets (#40)
redmedical/hvstr: docs/guide/custom-snipptes.md

# Custom Snippets

> Assumed that you need hvstr to generate some custom functionality for your elements. Therefore hvstr has the custom snippets feature.

Reviewer: Therefore heißt "deswegen" -> Kausalität :)

Muss mich damit aber gleich entschuldigen.

Ich kannte nur "therefore" und nicht "therefor" und habe es dir einmal verbessert.

Habe es eben extra gegooglet und "therefor" heißt "dafür" also hast du es korrekt verwendet.
Scheint aber "juristische" Sprache zu sein.

Gibt also therefore und therefor.

Finde "For that" oder "For this" gebräuchlicher aber werde dir "therefor" nicht mehr verbwessern.

It allows you to add code to the page-object generation process, which will be added, under a given condition.

There are three different types of custom snippets. Each type will be generated at a different location in the code:

* ### CustomSnippets For Getter functions

> This type of custom snippets will be placed after each getter function of an element. It enables you to add custom function, which for example can toggle an element or selecting an sub element. For this type you will need to generate a function header, because your code will be placed inside the page-object class.

Reviewer: Hat ein Element mehrere getter? Wird das custom snippet mehrmals generiert?
Falls nein -> will be placed after the getter function...

Author: the getterFunction is also an custom Snippet, which is always present.

  }
```

* ### CustomSnippets For FillForm

> In this type you can extend the [FillForm](./add-meethodes.md#addfillform) methode, to let you fill out your custom form elements.

Reviewer: Wo wird es das custom snippet in fillform sein?
Anfang? Mitte? Ende?

Reviewer: Gleiches wäre bei clearForm interessant

Author: all custom snippets will be inserted sorted by the elements. The order is not relevant...

Reviewer: Ok. My thought was just that maybe somebody wants to change the order. For example if the last input field is a field that only triggers validation if you clicked somewhere else. So he would not want to fill it out at the end.
But like Michel said, usually you fill out in order and those are edge cases.
Just wanted to add that.

Reviewer: So you could just add:

In the fillForm method the methods to fill the different elements are ordered the same way the elements are ordered in the DOM.

Author: that is not necessarily true. The order can be out of order, for example if an element is displayed later.

> The fillForm methode requires a data parameter. You can define the type in the custom snippet.

Reviewer: The fillForm methode automatically has a "data" input parameter. You can define its properties in the custom snippet.

Stimmt das so?
Falls ja finde ich es so noch verständlicher.

#### Example
this custom snippet:
```ts
    pageObjectBuilder.customSnippets.addForFillForm({

>         condition: (element) => element.type === 'INPUT', // Add this custom snippet for all elements of type input

Reviewer: Vielleicht irgendwo schreiben, dass man hier alles groß schreiben MUSS.
Ist doch so oder?
Oder wird 'input' auch akzeptiert?

        callBack: async (
            element: E2eElement,
            codeBuilder: QueuedCodeBuilder,
            options: IPageObjectBuilderOptions,
        ) => {
            codeBuilder

>                 .addLine(`if (data.${Utils.firstCharToLowerCase(element.id)}) {`) // make sure, if the entity on the data object is set

Reviewer: ```suggestion
                .addLine(`if (data.${Utils.firstCharToLowerCase(element.id)}) {`) // make sure, that the property on the data object is set
```

Reviewer: if = falls

garantiere, falls die entity auf dem data object gesetzt ist

But it's not important to me

                .increaseDepth()
                .addLine(`await this.get${element.id}().sendKeys(data.${Utils.firstCharToLowerCase(element.id)});`)
                .decreaseDepth()
                .addLine('}');
        },
        type: 'string', // fillForm requires type string for this field
    });
```
will look like this in your page-object:

```ts
 async fillForm(
    data: {
      // [...]
      nameInput?: string;
      // [...]
    },
  ) {
    // [...]
    if (data.nameInput) {
      await this.getNameInput().sendKeys(data.nameInput);
    }
    // [...]
  }
```

* ### CustomSnippets For ClearForm

> In this type you can extend the [ClearForm](./add-methods.md#addfillform) methode, to let you clear your form elements.

Reviewer: method
